### PR TITLE
Hooking up

### PR DESCRIPTION
**Pull request by plor**
we should probably actually start merging this stuff to main.

### CODE REVIEW

File: mist-invoice-contracts/src/MistInvoiceEscrowWrapper.sol [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+            _mistData.merkleRoot,
+            _mistData.clientRandom,
+            _mistData.providerRandom,
+            _mistData.clientKey,
+            _mistData.providerKey
+        );
+        
+        // create new invoice
+        ISmartInvoiceFactory factory = ISmartInvoiceFactory(INVOICE_FACTORY);
+        invoice = factory.create(address(mistPool), _amounts, _data, _type);
```
Comment (geovgy) [inline mist-invoice-contracts/src/MistInvoiceEscrowWrapper.sol:56]: Should set `address(this)` as client, not `address(mistPool)`

File: mist-invoice-contracts/src/MistInvoiceEscrowWrapper.sol [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+        // TODO verify proof
+        // TODO call lock on _invoiceAddr
+        ISmartInvoiceEscrow escrow = ISmartInvoiceEscrow(_invoiceAddr);
+        escrow.lock(_details);
+    }
+
+    // TODO other resolve arguments needed
+    function resolve(address _invoiceAddr, uint256 _clientAward, uint256 _providerAward, bytes32 _details) external {
+        require(_invoiceAddr != address(0), "valid invoice required");
+        // TODO check arbitration address msg.sender
```
Comment (geovgy) [inline mist-invoice-contracts/src/MistInvoiceEscrowWrapper.sol:96]: Is this TODO needed if we are doing delegate call?

File: mist-invoice-contracts/src/MistInvoiceEscrowWrapper.sol [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+    // TODO how many bytes?
+    struct MistSecret {
+        bytes merkleRoot;
+        bytes encClientRandInt;
+        bytes encProviderRandInt;
+        bytes encClientKey;
+        bytes encProviderKey;
+    }
+
+    struct MistData {
```
Comment (geovgy) [inline mist-invoice-contracts/src/MistInvoiceEscrowWrapper.sol:25]: I don't think this struct is necessary.

File: mist-invoice-contracts/src/MistInvoiceEscrowWrapper.sol [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+    mapping(address => MistSecret) mistSecrets;
+
+    // TODO how many bytes?
+    struct MistSecret {
+        bytes merkleRoot;
+        bytes encClientRandInt;
+        bytes encProviderRandInt;
+        bytes encClientKey;
+        bytes encProviderKey;
+    }
```
Comment (geovgy) [inline mist-invoice-contracts/src/MistInvoiceEscrowWrapper.sol:23]: We could reduce the number of items in this struct to:
```solidity=
{
     bytes32 merkleRoot; // or uint256 type
     uint256 providerHash; // hash of the random int, provider address and nullifying key to be used in MIST's deposit function as the receiverHash in PreCommitment struct
     uint256 clientHash; // same thing for client
     bytes encData; // encode both encrypted keys and the encrypted random uints in a single bytes array.
}
```

**Review by geovgy — APPROVED** [review]
Solid so far. There are some areas where privacy can be leaked. If we remove the MistData struct, we should avoid that issue. Plus saves some gas.

General rule of thumb is if its MIST related, it should be hashed or encrypted before touching onchain.

File: mist-invoice-contracts/src/MistInvoiceEscrowWrapper.sol (hunk 2 above)
Comment (plor) [reply mist-invoice-contracts/src/MistInvoiceEscrowWrapper.sol:96]: I don't think so, you're right.

**Pull request merged**